File: MarvinPlugin/Source/MarvinPlugin/MarvinPlugin.m
//
// MarvinPlugin.m
// Marvin for Xcode
//
// Created by Christoffer Winterkvist on 17/10/14.
// Copyright (c) 2014 zenangst The MIT License.
//

#import "MarvinPlugin.h"
#import "XcodeManager.h"
#import "NSDocument+ZENProperSave.h"
#import "MarvinSettingsWindowController.h"

static MarvinPlugin *marvinPlugin;

@interface MarvinPlugin ()

@property (nonatomic, strong) XcodeManager *xcodeManager;
@property MarvinSettingsWindowController *settingsWindowController;


@end

@implementation MarvinPlugin

+ (void)pluginDidLoad:(NSBundle *)plugin {
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        marvinPlugin = [[self alloc] init];
        if ( !(marvinPlugin.settingsWindowController = [[MarvinSettingsWindowController alloc] initWithBundle:plugin]) ) {
            NSLog( @"MarvinPlugin: nib not loaded exiting" );
            return;
        }
    });
}

- (void)dealloc {
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}

- (instancetype)init {
    self = [super init];
    if (!self) return nil;

    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(applicationDidFinishLaunching:)
                                                 name:NSApplicationDidFinishLaunchingNotification
                                               object:nil];

    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(properSave)
                                                 name:@"Save properly"
                                               object:nil];

    return self;
}

- (void)applicationDidFinishLaunching:(NSNotification *)notification {
    NSMenuItem *editMenuItem = [[NSApp mainMenu] itemWithTitle:@"Edit"];

    if (editMenuItem) {
        NSMenu *marvinMenu = [[NSMenu alloc] initWithTitle:@"Marvin"];

        [[editMenuItem submenu] addItem:[NSMenuItem separatorItem]];

        [marvinMenu addItem:({
            NSMenuItem *menuItem = [[NSMenuItem alloc] initWithTitle:@"Settings"
                                                              action:@selector(settingsMenuItemSelected:)
                                                       keyEquivalent:@""];
            menuItem.target = self;
            menuItem;
        })];

        [marvinMenu addItem:[NSMenuItem separatorItem]];

        [marvinMenu addItem:({
            NSMenuItem *menuItem = [[NSMenuItem alloc] initWithTitle:@"Delete Line"
                                                              action:@selector(deleteLineAction)
                                                       keyEquivalent:@""];
            menuItem.target = self;
            menuItem;
        })];

        [marvinMenu addItem:({
            NSMenuItem *menuItem = [[NSMenuItem alloc] initWithTitle:@"Duplicate Line"
                                                              action:@selector(duplicateLineAction)
                                                       keyEquivalent:@""];
            menuItem.target = self;
            menuItem;
        })];

        [marvinMenu addItem:({
            NSMenuItem *menuItem = [[NSMenuItem alloc] initWithTitle:@"Join Line"
                                                              action:@selector(joinLineAction)
                                                       keyEquivalent:@""];
            menuItem.target = self;
            menuItem;
        })];

        [marvinMenu addItem:({
            NSMenuItem *menuItem = [[NSMenuItem alloc] initWithTitle:@"Move To EOL and Insert LF"
                                                              action:@selector(moveToEOLAndInsertLFAction)
                                                       keyEquivalent:@""];
            menuItem.target = self;
            menuItem;
        })];

        [marvinMenu addItem:({
            NSMenuItem *menuItem = [[NSMenuItem alloc] initWithTitle:@"Select Current Word"
                                                              action:@selector(selectWordAction)
                                                       keyEquivalent:@""];
            menuItem.target = self;
            menuItem;
        })];

        [marvinMenu addItem:({
            NSMenuItem *menuItem = [[NSMenuItem alloc] initWithTitle:@"Select Line Contents"
                                                              action:@selector(selectLineContentsAction)
                                                       keyEquivalent:@""];
            menuItem.target = self;
            menuItem;
        })];

        [marvinMenu addItem:({
            NSMenuItem *menuItem = [[NSMenuItem alloc] initWithTitle:@"Select Next Word"
                                                              action:@selector(selectNextWordAction)
                                                       keyEquivalent:@""];
            menuItem.target = self;
            menuItem.keyEquivalentModifierMask = NSControlKeyMask;
            menuItem;
        })];

        [marvinMenu addItem:({
            NSMenuItem *menuItem = [[NSMenuItem alloc] initWithTitle:@"Select Previous Word"
                                                              action:@selector(selectPreviousWordAction)
                                                       keyEquivalent:@""];
            menuItem.target = self;
            menuItem;
        })];

        [marvinMenu addItem:({
            NSMenuItem *menuItem = [[NSMenuItem alloc] initWithTitle:@"Select Word Above"
                                                              action:@selector(selectWordAboveAction)
                                                       keyEquivalent:@""];
            menuItem.target = self;
            menuItem.keyEquivalentModifierMask = NSControlKeyMask;
            menuItem;
        })];

        [marvinMenu addItem:({
            NSMenuItem *menuItem = [[NSMenuItem alloc] initWithTitle:@"Select Word Below"
                                                              action:@selector(selectWordBelowAction)
                                                       keyEquivalent:@""];
            menuItem.target = self;
            menuItem.keyEquivalentModifierMask = NSControlKeyMask;
            menuItem;
        })];

        [marvinMenu addItem:({
            NSMenuItem *menuItem = [[NSMenuItem alloc] initWithTitle:@"Sort Lines"
                                                              action:@selector(sortLines)
                                                       keyEquivalent:@""];
            menuItem.target = self;
            menuItem;
        })];

        NSString *versionString = [[NSBundle bundleForClass:[self class]] objectForInfoDictionaryKey:@"CFBundleVersion"];
        NSMenuItem *marvinMenuItem = [[NSMenuItem alloc] initWithTitle:[NSString stringWithFormat:@"Marvin (%@)", versionString]
                                                                action:nil
                                                         keyEquivalent:@""];
        marvinMenuItem.submenu = marvinMenu;

        [[editMenuItem submenu] addItem:marvinMenuItem];
    }
}

#pragma mark - Getters

- (XcodeManager *)xcodeManager {
    if (_xcodeManager) return _xcodeManager;

    _xcodeManager = [[XcodeManager alloc] init];

    return _xcodeManager;
}

- (BOOL)validResponder {
    NSResponder *firstResponder = [[NSApp keyWindow] firstResponder];
    NSString *responderClass = NSStringFromClass(firstResponder.class);
    NSArray *validClasses = @[@"DVTSourceTextView", @"IDEPlaygroundTextView"];

    return ([validClasses containsObject:responderClass]);
}

#pragma mark - Actions

- (void)settingsMenuItemSelected:(id)sender {
    [marvinPlugin.settingsWindowController showWindow:self];
}

- (void)selectLineContentsAction {
    if ([self validResponder]) {
        self.xcodeManager.selectedRange = self.xcodeManager.lineContentsRange;
    }
}

- (void)selectWordAction {
    if ([self validResponder]) {
        self.xcodeManager.selectedRange = self.xcodeManager.currentWordRange;
    }
}

- (void)selectWordAboveAction {
    if ([self validResponder]) {
        NSCharacterSet *validSet = [NSCharacterSet characterSetWithCharactersInString:kMarvinValidSetWordString];
        NSRange currentRange = [self.xcodeManager selectedRange];
        unichar characterAtCursorStart = [[self.xcodeManager contents]
                                          characterAtIndex:currentRange.location];
        unichar characterAtCursorEnd = [[self.xcodeManager contents]
                                        characterAtIndex:currentRange.location-1];

        if (![self.xcodeManager selectedRange].length &&
            [validSet characterIsMember:characterAtCursorStart]) {
            [self selectWordAction];
        } else if (![self.xcodeManager selectedRange].length &&
                   [validSet characterIsMember:characterAtCursorEnd]) {
            [self selectPreviousWordAction];
        } else {
            [self performKeyboardEvent:126];

            dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.025 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{

                NSRange currentRange = [self.xcodeManager selectedRange];
                unichar characterAtCursorStart = [[self.xcodeManager contents] characterAtIndex:currentRange.location];

                if ([validSet characterIsMember:characterAtCursorStart]) {
                    [self selectWordAction];
                } else {
                    [self selectPreviousWordAction];
                }
            });
        }
    }
}

- (void)selectWordBelowAction {
    if ([self validResponder]) {
        [self performKeyboardEvent:125];

        dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.025 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
            [self selectWordAction];
        });
    }
}

- (void)selectPreviousWordAction {
    if ([self validResponder]) {
        self.xcodeManager.selectedRange = self.xcodeManager.previousWordRange;
        self.xcodeManager.selectedRange = self.xcodeManager.currentWordRange;
    }
}

- (void)selectNextWordAction {
    if ([self validResponder]) {
        [self selectWordAction];
    }
}

- (void)deleteLineAction {
    if ([self validResponder]) {
        [self.xcodeManager replaceCharactersInRange:self.xcodeManager.lineRange
                                         withString:@""];
    }
}

- (void)duplicateLineAction {
    if ([self validResponder]) {
        NSRange range = [self.xcodeManager lineRange];
        NSMutableString *string = [[self.xcodeManager contentsOfRange:range] mutableCopy];
        NSRange duplicateRange = NSMakeRange(range.location+range.length, 0);
        NSUInteger offset = 0;

        if (duplicateRange.location >= [[self.xcodeManager contents] length]) {
            [string insertString:@"\n" atIndex:0];
            offset += 1;
        }

        [self.xcodeManager replaceCharactersInRange:duplicateRange
                                         withString:[string copy]];
        NSRange selectRange = NSMakeRange(duplicateRange.location + duplicateRange.length + string.length - 1 + offset, 0);
        [self.xcodeManager setSelectedRange:selectRange];
    }
}

- (void)joinLineAction {
    if ([self validResponder]) {
        if ([self.xcodeManager lineContentsRange].length > 0) {
            [self.xcodeManager replaceCharactersInRange:self.xcodeManager.joinRange
                                             withString:@" "];
        } else {
            [self.xcodeManager replaceCharactersInRange:self.xcodeManager.joinRange
                                             withString:@""];
        }
    }
}

- (void)moveToEOLAndInsertLFAction {
    NSRange lineContentsRange = self.xcodeManager.lineContentsRange;
    NSRange lineRange = [self.xcodeManager lineRange];

    if (lineContentsRange.location == NSNotFound) {
        lineContentsRange.location = self.xcodeManager.contents.length;
        lineContentsRange.length = 0;

        [self.xcodeManager replaceCharactersInRange:lineContentsRange
                                         withString:@"\n"];
        self.xcodeManager.selectedRange = NSMakeRange(lineContentsRange.location+1, 0);

        return;
    }

    unsigned long endOfLine = (unsigned long)lineContentsRange.location+(unsigned long)lineContentsRange.length;

    NSString *spacing = [[self.xcodeManager contents] substringWithRange:NSMakeRange(lineRange.location, lineContentsRange.location - lineRange.location)];

    unichar lastCharacterInLine = [[self.xcodeManager contents] characterAtIndex:lineContentsRange.location+lineContentsRange.length-1];
    int ascii = lastCharacterInLine;

    NSMutableString *additionalSpacing = [NSMutableString new];
    if (ascii == 123) {
        for (int x = 0; x < 0; x++) {
            [additionalSpacing appendString:@" "];
        }
    }

    [self.xcodeManager replaceCharactersInRange:NSMakeRange(endOfLine,0)
                                     withString:[NSString stringWithFormat:@"\n%@%@", spacing, [additionalSpacing copy]]];
    [self.xcodeManager setSelectedRange:NSMakeRange(endOfLine+1+spacing.length+additionalSpacing.length, 0)];
}

- (void)properSave {
    [self removeTrailingWhitespace:^{
        [self addNewlineAtEOF];
        [self.xcodeManager save];
    }];
}

- (void)sortLines {
    NSRange lineRange = [self.xcodeManager lineRange];
    NSString *selectedContent = [self.xcodeManager contentsOfRange:lineRange];
    NSArray *lines = [selectedContent componentsSeparatedByString:@"\n"];
    NSArray *sortedLinesArray = [lines sortedArrayUsingSelector:@selector(localizedCaseInsensitiveCompare:)];
    NSString *sortedLinesString = [sortedLinesArray componentsJoinedByString:@"\n"];

    BOOL shouldSortDescending = ([[selectedContent substringToIndex:selectedContent.length-1] isEqualToString:[sortedLinesString substringFromIndex:1]]);
    if (shouldSortDescending) {
        NSSortDescriptor *sortOrder = [NSSortDescriptor sortDescriptorWithKey:@"self"
                                                                    ascending:NO];
        sortedLinesArray = [sortedLinesArray sortedArrayUsingDescriptors:[NSArray arrayWithObject:sortOrder]];
        sortedLinesString = [sortedLinesArray componentsJoinedByString:@"\n"];
        [self.xcodeManager replaceCharactersInRange:lineRange
                                         withString:sortedLinesString];
    } else {
        if (lineRange.location > 0) { lineRange.location -= 1; }
        if ((lineRange.location + lineRange.length) >= [[self.xcodeManager contents] length]) {
            lineRange = NSMakeRange(0, [[self.xcodeManager contents] length]);
        }
        [self.xcodeManager replaceCharactersInRange:lineRange
                                         withString:sortedLinesString];
    }
}

#pragma mark - Private methods

- (void)addNewlineAtEOF {
    if ([self validResponder]) {

        NSString *documentText = self.xcodeManager.contents;

        if (self.xcodeManager.contents.length) {
            int eof = [documentText characterAtIndex:[documentText length]-1];
            int lastASCII = [documentText characterAtIndex:[documentText length]-2];

            if (lastASCII != 100 && eof != 10) {
                NSRange selectedRange = self.xcodeManager.selectedRange;
                NSRange replaceRange = NSMakeRange(self.xcodeManager.contents.length, 0);
                NSString *replaceString = [NSString stringWithFormat:@"%c", 10];

                [self.xcodeManager replaceCharactersInRange:replaceRange
                                                 withString:replaceString];
                self.xcodeManager.selectedRange = selectedRange;
            }
        }
    }
}

- (void)removeTrailingWhitespace:(void (^)())block {
    if (![self validResponder]) {
        block();
        return;
    }

    NSString *key = @"MarvinRemoveWhitespace";
    BOOL shouldRemoveWhitespace = [[NSUserDefaults standardUserDefaults] boolForKey:key];

    if (shouldRemoveWhitespace) {
        block();
        return;
    }


    if (!marvinPlugin.settingsWindowController.shouldRemoveWhitespace.state) {
        block();
        return;
    }

    NSError *error = nil;
    NSRegularExpression *regex = [NSRegularExpression regularExpressionWithPattern:@"([ \t]+)\r?\n"
                                                                           options:NSRegularExpressionCaseInsensitive
                                                                             error:&error];

    if (error) {
        NSLog(@"Couldn't create regex with given string and options");
    }

    NSString *string = self.xcodeManager.contents;
    NSRange currentRange = self.xcodeManager.selectedRange;
    NSMutableArray *ranges = [NSMutableArray new];

    [regex enumerateMatchesInString:string options:NSMatchingReportProgress
                              range:NSMakeRange(0,[string length])
                         usingBlock:^(NSTextCheckingResult *result, NSMatchingFlags flags, BOOL *stop) {

                             if (result && !NSLocationInRange(currentRange.location, result.range)) {
                                 [ranges addObject:result];
                             }

                         }];

    if (![ranges count]) {
        block();
        return;
    }

    NSEnumerator *enumerator = [ranges reverseObjectEnumerator];

    dispatch_async(dispatch_get_main_queue(),^{
        for (NSTextCheckingResult *textResult in enumerator) {
            NSRange range = textResult.range;
            range.length -= 1;
            [self.xcodeManager replaceCharactersInRange:range withString:@""];
        }
        block();
    });
}

- (void)performKeyboardEvent:(CGKeyCode)virtualKey {
    CGEventRef event = CGEventCreateKeyboardEvent(NULL, virtualKey, true);
    CGEventSetFlags(event, 0);
    CGEventPost(kCGHIDEventTap, event);
    CFRelease(event);
}

@end

File: MarvinPlugIn/Source/MarvinPlugIn/MarvinPlugin.swift
import Cocoa

var marvinPlugin: MarvinPlugin? = nil

extension NSObject {

  class func pluginDidLoad(_ bundle: Bundle) {
    guard let appName = Bundle.main.infoDictionary?["CFBundleName"] as? String
      , marvinPlugin == nil && appName == "Xcode" else {
        return
    }

    marvinPlugin = MarvinPlugin()
    marvinPlugin?.settingsController = MarvinSettingsWindowController(bundle: bundle)
    SaveSwizzler.swizzle()
  }
}

class MarvinPlugin: NSObject {

  lazy var xcode = XcodeManager()
  var settingsController: MarvinSettingsWindowController?

  deinit {
    NotificationCenter.default.removeObserver(self)
  }

  override init() {
    super.init()

    NotificationCenter.default.addObserver(self, selector: #selector(NSApplicationDelegate.applicationDidFinishLaunching(_:)), name: NSNotification.Name.NSApplicationDidFinishLaunching, object: nil)

    NotificationCenter.default.addObserver(self, selector: #selector(MarvinPlugin.properSave), name: NSNotification.Name(rawValue: "Save properly"), object: nil)
  }

  func applicationDidFinishLaunching(_ notification: Notification) {
    guard let mainMenu = NSApp.mainMenu else { return }
    let editMenuItem = mainMenu.item(withTitle: "Edit")

    if let editMenuItem = editMenuItem, let submenu = editMenuItem.submenu {
      let marvinMenu = NSMenu.init(title: "Marvin")
      var items = [NSMenuItem]()

      items.append(NSMenuItem.init(title: "Settings", action: #selector(MarvinPlugin.settingsMenuItemSelected), keyEquivalent: ""))
      items.append(NSMenuItem.separator())
      items.append(NSMenuItem.init(title: "Delete Line", action: #selector(MarvinPlugin.deleteLineAction), keyEquivalent: ""))
      items.append(NSMenuItem.init(title: "Duplicate Line", action: #selector(MarvinPlugin.duplicateLineAction), keyEquivalent: ""))
      items.append(NSMenuItem.init(title: "Join Line", action: #selector(MarvinPlugin.joinLineAction), keyEquivalent: ""))
      items.append(NSMenuItem.init(title: "Move To EOL and Insert LF", action: #selector(MarvinPlugin.moveToEOLAndInsertLFAction), keyEquivalent: ""))
      items.append(NSMenuItem.init(title: "Select Current Word", action: #selector(MarvinPlugin.selectWordAction), keyEquivalent: ""))
      items.append(NSMenuItem.init(title: "Select Line Contents", action: #selector(MarvinPlugin.selectLineContentsAction), keyEquivalent: ""))
      items.append(NSMenuItem.init(title: "Select Next Word", action: #selector(MarvinPlugin.selectNextWordAction), keyEquivalent: ""))
      items.append(NSMenuItem.init(title: "Select Previous Word", action: #selector(MarvinPlugin.selectPreviousWordAction), keyEquivalent: ""))
      items.append(NSMenuItem.init(title: "Select Word Above", action: #selector(MarvinPlugin.selectWordAboveAction), keyEquivalent: ""))
      items.append(NSMenuItem.init(title: "Select Word Below", action: #selector(MarvinPlugin.selectWordBelowAction), keyEquivalent: ""))
      items.append(NSMenuItem.init(title: "Sort Lines", action: #selector(MarvinPlugin.sortLines), keyEquivalent: ""))

      items.forEach { $0.target = self; marvinMenu.addItem($0) }

      if let infoDictionary = Bundle(for: type(of: self)).infoDictionary,
        let version = infoDictionary["CFBundleVersion"] {
          let marvinMenuItem = NSMenuItem.init(title: "Marvin \(version)", action: nil, keyEquivalent: "")
          marvinMenuItem.submenu = marvinMenu

          submenu.addItem(NSMenuItem.separator())
          submenu.addItem(marvinMenuItem)
      }
    }
  }

  func validResponder() -> Bool {
    guard let window = NSApp.keyWindow else { return false }
    let firstResponder = window.firstResponder
    let responderClass = NSStringFromClass(type(of: firstResponder))

    return ["NSKVONotifying_DVTSourceTextView", "NSKVONotifying_IDEPlaygroundTextView"].contains(responderClass) && xcode.documentLength() > 1
  }

  func settingsMenuItemSelected() {
    marvinPlugin?.settingsController?.openWindow()
  }

  func selectLineContentsAction() {
    guard validResponder() else { return }
    xcode.selectedRange = xcode.lineContentsRange()
  }

  func selectWordAction() {
    guard validResponder() else { return }
    xcode.selectedRange = xcode.currentWordRange()
  }

  func selectWordAboveAction() {
    guard validResponder() && xcode.selectedRange.location > 0 else { return }

    let validSet = CharacterSet(charactersIn: "0123456789ABCDEFGHIJKOLMNOPQRSTUVWXYZÅÄÆÖØabcdefghijkolmnopqrstuvwxyzåäæöø_")
    var currentRange = xcode.selectedRange

    if currentRange.location >= xcode.contents().characters.count {
      currentRange.location -= 1
    }

    let characterAtCursorStart: Character = xcode.contents()[xcode.contents().characters.index(xcode.contents().startIndex, offsetBy: currentRange.location)]
    let characterAtCursorEnd: Character = xcode.contents()[xcode.contents().characters.index(xcode.contents().startIndex, offsetBy: currentRange.location-1)]

    if xcode.selectedRange.length == 0 && isChar(characterAtCursorStart, inSet: validSet) {
      selectWordAction()
    } else if xcode.selectedRange.length == 0 && isChar(characterAtCursorEnd, inSet: validSet) {
      selectPreviousWordAction()
    } else {
      perform(keyboardEvent: 126)

      let delayTime = DispatchTime.now() + 0.025
      DispatchQueue.main.asyncAfter(deadline: delayTime) { [unowned self] in
        let currentRange = self.xcode.selectedRange

        let characterAtCursorStart: Character = self.xcode.contents()[self.xcode.contents().characters.index(self.xcode.contents().startIndex, offsetBy: currentRange.location)]

        if self.isChar(characterAtCursorStart, inSet: validSet) {
          self.selectWordAction()
        } else {
          self.selectPreviousWordAction()
        }
      }
    }
  }

  func selectWordBelowAction() {
    guard validResponder() else { return }

    perform(keyboardEvent: 125)

    let delayTime = DispatchTime.now() + 0.025
    DispatchQueue.main.asyncAfter(deadline: delayTime) { [unowned self] in
      self.selectWordAction()
    }
  }

  func selectPreviousWordAction() {
    guard validResponder() else { return }
    xcode.selectedRange = xcode.previousWordRange()
    xcode.selectedRange = xcode.currentWordRange()
  }

  func selectNextWordAction() {
    guard validResponder() else { return }
    selectWordAction()
  }

  func deleteLineAction() {
    guard validResponder() else { return }
    xcode.replaceCharactersInRange(xcode.lineRange(), withString: "")
  }

  func duplicateLineAction() {
    guard validResponder() else { return }

    let range = xcode.lineRange()
    var string = self.xcode.contentsOfRange(range)
    let duplicateRange = NSMakeRange(range.location+range.length, 0)
    var offset = 0

    if duplicateRange.location >= xcode.contents().characters.count &&
      xcode.selectedRange.location == xcode.contents().characters.count {
        string = "\n\(string)"
        offset = 1
    }

    xcode.replaceCharactersInRange(duplicateRange, withString: string)
    xcode.selectedRange = NSMakeRange(duplicateRange.location + duplicateRange.length + string.characters.count - 1 + offset, 0)
  }

  func joinLineAction() {
    guard validResponder() else { return }

    let range = xcode.joinRange()

    guard range.location != NSNotFound &&
      range.location + range.length < xcode.contents().characters.count
      else { return }

    xcode.replaceCharactersInRange(range,
      withString: xcode.lineContentsRange().length > 0 ? " " : "")
  }

  func moveToEOLAndInsertLFAction() {
    guard validResponder() else { return }

    let lineRange = xcode.lineRange()
    let endOfLine = lineRange.location + lineRange.length - 1
    let currentLine = (xcode.contents() as NSString).substring(with: lineRange)
    let trimmedString = currentLine.trimmingCharacters(in: CharacterSet.whitespaces)
    let spacing = currentLine.replacingOccurrences(of: trimmedString, with: "")

    xcode.replaceCharactersInRange(NSMakeRange(endOfLine, 0), withString: "\n\(spacing)")
    xcode.selectedRange = NSMakeRange(endOfLine + spacing.characters.count + 1, 0)
  }

  func sortLines() {
    guard validResponder() else { return }

    var lineRange = xcode.lineRange()
    lineRange.length -= 1
    let selectedContent = xcode.contentsOfRange(lineRange)
    let lines = selectedContent.components(separatedBy: "\n")

    var sortedLines = lines.sorted { $0 > $1 }
    var sortedLinesString = (sortedLines as NSArray).componentsJoined(by: "\n")

    let shouldSortDescending = (selectedContent as NSString).trimmingCharacters(in: CharacterSet.whitespacesAndNewlines) == (sortedLinesString as NSString).substring(from: 1).trimmingCharacters(in: CharacterSet.whitespacesAndNewlines)

    if shouldSortDescending {
      sortedLines = lines.sorted { $0 < $1 }
      sortedLinesString = (sortedLines as NSArray).componentsJoined(by: "\n")
    }

    xcode.replaceCharactersInRange(lineRange, withString: sortedLinesString)
    xcode.selectedRange = lineRange
  }

  func properSave() {
    removeTrailingWhitespace { [unowned self] in
      self.addNewlineAtEOF()
      self.xcode.save()
    }
  }

  // MARK: - Private methods

  fileprivate func addNewlineAtEOF() {
    guard validResponder() else { return }

    if let eof = xcode.contents().characters.last
      , eof != "\n" {
        let selectedRange = xcode.selectedRange
        let replaceRange = NSMakeRange(xcode.contents().characters.count, 0)
        let replaceString = "\n"

        xcode.replaceCharactersInRange(replaceRange, withString: replaceString)
        xcode.selectedRange = selectedRange
    }
  }

  fileprivate func removeTrailingWhitespace(_ closure: @escaping () -> Void) {
    guard validResponder() else { closure(); return }

    let key = "MarvinRemoveWhitespace"
    let shouldRemoveWhitespace = UserDefaults.standard.bool(forKey: key)

    if !shouldRemoveWhitespace {
      closure()
      return
    }

    let regex = try! NSRegularExpression(pattern: "([ \t]+)\r?\n", options: .caseInsensitive)
    let currentRange = xcode.selectedRange
    let string = xcode.contents()

    var results = [NSTextCheckingResult]()

    regex.enumerateMatches(in: string, options: .reportProgress, range: NSMakeRange(0, string.characters.count)) { (result, flags, stop) -> Void in
      if let result = result , !NSLocationInRange(currentRange.location, result.range) {
        results.append(result)
      }
    }

    if results.isEmpty { closure(); return }

    let enumerator = results.reversed()

    DispatchQueue.main.async { [unowned self] in
      enumerator.forEach { textResult in
        var range = textResult.range
        range.length -= 1
        self.xcode.replaceCharactersInRange(range, withString: "")
      }
      closure()
    }
  }

  fileprivate func perform(keyboardEvent virtualKey: CGKeyCode) {
    let event = CGEvent(keyboardEventSource: nil, virtualKey: virtualKey, keyDown: true)
    event?.flags = CGEventFlags(rawValue: 0)
    event?.post(tap: .cghidEventTap)
  }

  fileprivate func isChar(_ char: Character, inSet set: CharacterSet) -> Bool {
    var found = false
    for ch in String(char).utf16 {
      if set.contains(UnicodeScalar(ch)!) { found = true; break }
    }
    return found
  }
}
